### ISSUE TITLE
Using target specification causes a panic

### ISSUE THREAD

**Issue by Tobba** [rust-lang/cargo#1159]
In some unwrap I can't track down.

This only occurs if rustc would consider the target specification "valid", except it seemingly never actually gets to running rustc, thus this makes no sense whatsoever.

```
$ cat test.json
{
        ...
}
$ cargo build --target test
thread '<main>' panicked at 'called `Option::unwrap()` on a `None` value', C:\bot\slave\nightly-dist-rustc-win-64\build\src\libcore\option.rs:361
$ cat test2.json
invalid
$ cargo build --target test2
An unknown error occurred
```
(In this case the unknown error is the target spec being invalid, which it'll tell you with `--verbose`)

**Comment by DiamondLovesYou** [comment]
This isn't limited to target specifications: when I try to target ```le32-unknown-nacl```, which doesn't use a target spec, I get (I believe) the same ```unwrap()``` panic.

**Comment by Tobba** [comment]
I managed to track it down to here
https://github.com/rust-lang/cargo/blob/4a0154d4201d99bb7b091764a32c062edef53ff4/src/cargo/ops/cargo_rustc/context.rs#L101-L116
As you can see that code is pretty much complete bonkers. And it turns out someone adjusted the warning message slightly, breaking it completely and triggering this unwrap.

I'd fix it myself if I could get cargo to build on the latest Rust. But currently that doesn't work either and will require updates to nearly every library cargo uses.

**Comment by DiamondLovesYou** [comment]
I've almost completed an upgrade to master(ish) (a ```cargo update```
almost fixes everything anyway); I can push to my fork so you can test if
you'd like (or you can push your changes).

**Comment by Tobba** [comment]
The warnings strings now seem to be
```
dropping unsupported crate type `"dylib"`
dropping unsupported crate type `"bin"`
```

I'll push/test the change on that branch later today if you'd like

**Comment by DiamondLovesYou** [comment]
kk, cool. Cleaning up old Encoder/Decoder/etc usage now.

**Comment by alexcrichton** [comment]
Note that I've fixed this in #1161.

**Comment by DiamondLovesYou** [comment]
@alexcrichton Sweet, though I'll still need to continue fixin' shit so I can build cargo w/ your PR locally.
